proposals/0000-unlifted-datatypes.rst

```diff
@@ -0,0 +1,457 @@
+for unlifted functions as well as long as we don't allow functions without
+bindings.
+
+**-XStrict/-XStrictData** could implicitly turn all data declarations
+into ``unlifted`` ones. I see two potential problems:
+
+* If a data type is exported, it's now an unlifted type. That's a breaking change.
+* For data family instances, this is only possible if the parent data family
+  was kind polymorphic. Plus it's a strange thing to do change kinds of a
+  declaration just by switching on a language extension.
+
+So rather dreadfully, we probably shouldn't "augment" ``-XStrict``.
+
+**-XUnliftedNewtypes** introduces unlifted *newtypes*, but does so simply by
+inferring the kind of its single constructor's field type, no ``unlifted``
+needed. Now with the new ``unlifted`` keyword, we could potentially allow
+syntax like ``newtype unlifted Foo (a :: TYPE r) = Foo a``. What are its semantics?
+Can we still have ``Foo Int#``? That wouldn't exactly be ``UnliftedRep`` (which
+this proposal is all about), but the unlifted, unboxed runtime-rep ``IntRep``.
+Similarly, do we allow ``Foo Int``? That would be boxed and lifted, seemingly
+contradicting the declaration.
+
+So I suggest that we (somewhat ironically) *disallow* ``unlifted`` syntax for
+Newtype delcarations and instead suggest to activate ``-XUnliftedNewtypes``,
+which will automatically infer the generalised kind.
+
+**Lazy pattern matches** ``let ~t = ... in ...`` should not be allowed if ``t``
+is unlifted. That is exactly the behavior that is currently implemented in GHC.
+
+**Data con wrappers** of nullary constructors like
+``data unlifted T a where TInt :: T Int`` will become unlifted top-level
+bindings, which GHC currently forbids in source syntax. I *think* these will
+always be properly tagged, though, so it's just an implementation detail.
+The Core formalism should only be minimally affected by such unlifted bindings
+at the top-level. Either regard them as evaluated on start up or just allow
+normal forms (modulo lifted fields) See
+`#17521 <https://gitlab.haskell.org/ghc/ghc/issues/17521>`_.
+Another way forward would be to turn wrappers of nullary unlifted constructors
+into functions by adding a `Void#` argument.
+
+Costs and Drawbacks
+-------------------
+Thanks to previous work on unlifted types and ``-XUnliftedNewtypes``, this
+proposal seems rather easy to implement, with the majority of changes happening
+in the parser and type-checker. Notably the backend is not only affected **at
+all**. Very good cost to benefit ratio.
+
+As for the risk of making the language harder to learn: Beginners won't come in
+touch with unlifted datatypes at all. Unless they crave for better performance
+in a custom data structure, at which point I wouldn't consider them beginners
+anymore. There's precedent in going from unlifted to lifted by
+`Idris <http://docs.idris-lang.org/en/latest/tutorial/typesfuns.html>`_ with its
+``Lazy`` data type.
```

Comment:
Or Ocaml's `Lazy.t`.

Reply:
Fixed in https://github.com/ghc-proposals/ghc-proposals/pull/265/commits/356566d78fb08e6703c1c4acb808ddfdf1119efa.